dali/operators/numba_function/numba_func_gpu.cc

```diff
@@ -173,57 +200,52 @@ void NumbaFuncImpl<GPUBackend>::RunImpl(Workspace &ws) {
     auto& out = ws.Output<GPUBackend>(out_id);
     for (int i = 0; i < N; i++) {
-      out_ptrs[N * out_id + i] = reinterpret_cast<uint64_t>(out.raw_mutable_tensor(i));
+      out_arrays_[N * out_id + i].data = out.raw_mutable_tensor(i);
     }
   }
   for (size_t in_id = 0; in_id < in_types_.size(); in_id++) {
     auto& in = ws.Input<GPUBackend>(in_id);
     for (int i = 0; i < N; i++) {
-      in_ptrs[N * in_id + i] = reinterpret_cast<uint64_t>(in.raw_tensor(i));
+      in_arrays_[N * in_id + i].data = const_cast<void*>(in.raw_tensor(i));
     }
   }
 
   for (int i = 0; i < N; i++) {
     vector<void*> args;
+
+    for (size_t out_id = 0; out_id < out_types_.size(); out_id++) {
+      auto &dev_array = out_arrays_[out_id * N + i];
+      dev_array.PushArgs(args);
+    }
+
     for (size_t in_id = 0; in_id < in_types_.size(); in_id++) {
-      vector<void*> args_local = prepare_args(
-        in_memory_ptrs_[in_id],
-        in_sizes_[in_id],
-        &in_ptrs[N * in_id + i]);
-      args.insert(
-        args.end(),
-        make_move_iterator(args_local.begin()),
-        make_move_iterator(args_local.end()));
+      auto &dev_array = in_arrays_[in_id * N + i];
+      dev_array.PushArgs(args);
     }
 
-    for (size_t out_id = 0; out_id < out_types_.size(); out_id++) {
-      vector<void*> args_local = prepare_args(
-        out_memory_ptrs_[out_id],
-        out_sizes_[out_id],
-        &out_ptrs[N * out_id + i]);
-      args.insert(
-        args.end(),
-        make_move_iterator(args_local.begin()),
-        make_move_iterator(args_local.end()));
+    int blocks_per_sm_;
+    CUfunction cuFunc = reinterpret_cast<CUfunction>(run_fn_);
+    int blockDim = volume(threads_per_block_);
+    CUDA_CALL(cuOccupancyMaxActiveBlocksPerMultiprocessor(&blocks_per_sm_, cuFunc, blockDim, 0));
+    DALI_ENFORCE(blocks_per_sm_, "To many threads per block specified for the Numba provided GPU "
+                                 "kernel");
+    auto number_of_blocks = GetSmCount() * blocks_per_sm_;
+    if (number_of_blocks > volume(blocks_)) {
+      DALI_WARN(make_string("It is recommended that the grid volume: ", volume(blocks_),
+                            " for the Numba provided GPU kernel is at least: ",
+                            number_of_blocks));
     }
 
-    CUfunction cufunc = (CUfunction)run_fn_;
-    CUresult result = cuLaunchKernel(
-      cufunc,
-      blocks_[0], blocks_[1], blocks_[2],
-      threads_per_block_[0], threads_per_block_[1], threads_per_block_[2],
-      0,
-      ws.stream(),
-      static_cast<void**>(args.data()),
-      NULL);
-    cudaResultCheck(result);
+    CUDA_CALL(cuLaunchKernel(cuFunc,
+                             blocks_[0], blocks_[1], blocks_[2],
+                             threads_per_block_[0], threads_per_block_[1], threads_per_block_[2],
+                             0,
+                             ws.stream(),
+                             static_cast<void**>(args.data()),
```

Comment:
I don't think you need the cast when `args` is a `vector<void*>`
```suggestion
                             args.data(),
```

Reply:
done